sdk/src/main/java/com/google/cloud/dataflow/sdk/util/common/Counter.java

```diff
@@ -418,17 +430,44 @@ public boolean isCompatibleWith(Counter<?> that) {
 
   //////////////////////////////////////////////////////////////////////////////
 
-  /** The name of this counter. */
-  protected final String name;
-
   /** The kind of aggregation function to apply to this counter. */
   protected final AggregationKind kind;
 
   /** The commit state of this counter. **/
   protected final AtomicReference<CommitState> commitState;
 
+  /** A hashable name structure which will uniquely identify a counter. */
+  public interface Name {
+    public String name();
+    @Nullable public String systemName();
+    @Nullable public String originalName();
+  }
+
+  /** A StructuredName which uses some part of the NameContext. */
+  @AutoValue
+  public abstract static class StructuredName implements Name {
+    public static StructuredName createWithOriginalName(String name, NameContext context) {
+      return new AutoValue_Counter_StructuredName(name, null, checkNotNull(context.originalName()));
+    }
+
+    public static StructuredName createWithSystemName(String name, NameContext context) {
+      return new AutoValue_Counter_StructuredName(name, checkNotNull(context.systemName()), null);
+    }
+  }
+
+  /** An UnstructuerdName is the default, ie no NameContext set. */
+  @AutoValue
+  public abstract static class UnstructuredName implements Name {
```

Comment:
UnstructuredName = StructuredName right? Why not use the same thing, and just have a "createWithoutStructure(...)"?